### PR TITLE
Support standard UNION/INTERSECT syntax

### PR DESCRIPTION
**Pull request by dinodork**
The SQL standard allows parentheses around all query expressions (SELECTs) in a UNION. The Bison parser cannot currently handle queries like `(SELECT 1) UNION (SELECT 1)`, let alone `((SELECT 1) UNION (SELECT 1))`.

### CODE REVIEW

**Comment by dey4ss** [comment]
Just approved to run the actions, I will review your PR in the next week. I noticed that the gcc-6 stage fails because manually cloning does not work with forked repositories, I will see if I can provide a fix for that.

**Comment by dinodork** [comment]
> Just approved to run the actions, I will review your PR in the next week. I noticed that the gcc-6 stage fails because manually cloning does not work with forked repositories, I will see if I can provide a fix for that.

Thanks a lot. Since you're on it, I removed to last commit where I was trying to fix that issue, not doing a very good job at it.

**Comment by dey4ss** [comment]
The CI stage should work now if you merge the lastest update to master.

**Comment by dinodork** [comment]
> The CI stage should work now if you merge the lastest update to master.

Thanks. All green now!

File: Makefile [hunk 1]
```diff
@@ -70,7 +70,7 @@ $(SRCPARSER)/flex_lexer.o: $(SRCPARSER)/flex_lexer.cpp $(SRCPARSER)/bison_parser
 	$(CXX) $(LIB_CFLAGS) -c -o $@ $< -Wno-sign-compare -Wno-unneeded-internal-declaration -Wno-register
 
 %.o: %.cpp $(PARSER_CPP) $(LIB_H)
-	$(CXX) $(LIB_CFLAGS) -c -o $@ $<
+	$(CXX) $(LIB_CFLAGS) $(CXXFLAGS) -c -o $@ $<
```
Comment (dey4ss) [inline Makefile:73]: We don't use additonal flags in our base repo builds. Unless there's a reason to have them here, I'd ask you to remove them.

File: Makefile [hunk 2]
```diff
@@ -137,7 +137,7 @@ test: $(TEST_BUILD)
 
 $(TEST_BUILD): $(TEST_ALL) $(LIB_BUILD)
 	@mkdir -p $(BIN)/
-	$(CXX) $(TEST_CFLAGS) $(TEST_CPP) -o $(TEST_BUILD) -lsqlparser -lstdc++
+	$(CXX) $(CXXFLAGS) $(TEST_CFLAGS) $(TEST_CPP) -o $(TEST_BUILD) -lsqlparser -lstdc++
```
Comment (dey4ss) [inline Makefile:140]: (see above)

File: src/parser/bison_parser.y [hunk 3]
```diff
@@ -63,6 +63,8 @@
 // %output  "bison_parser.cpp"
 // %defines "bison_parser.h"
 
+%expect 0
```
Comment (dey4ss) [inline src/parser/bison_parser.y:66]: What does this do?

File: src/parser/bison_parser.y [hunk 4]
```diff
@@ -756,67 +759,137 @@ update_clause : IDENTIFIER '=' expr {
...
-  if ($$->setOperations == nullptr) {
-    $$->setOperations = new std::vector<SetOperation*>();
-  }
-  $$->setOperations->push_back($2);
-  $$->setOperations->back()->nestedSelectStatement = $3;
-  $$->setOperations->back()->resultOrder = $4;
-  $$->setOperations->back()->resultLimit = $5;
-  $$->lockings = $6;
-};
+select_statement:
```
Comment (dey4ss) [inline src/parser/bison_parser.y:762]: pick here: The indentation does not match the other code. Could ypu please fix that (e.g., by calling the `format.sh` script)?

File: src/parser/bison_parser.y [hunk 5]
```diff
@@ -756,67 +759,137 @@ update_clause : IDENTIFIER '=' expr {
...
+            }
+            else {
+              $2->setOperations->back()->resultOrder = $3;
+              $2->setOperations->back()->resultLimit = $4;
+            }
+
+            $$ = $2;
+          }
+
+/*
```
Comment (dey4ss) [inline src/parser/bison_parser.y:814]: knit-picky: We either use C-style comments with an asterisk `*` on each line or C++-style single line comments.

File: src/parser/bison_parser.y [hunk 6]
```diff
@@ -756,67 +759,137 @@ update_clause : IDENTIFIER '=' expr {
...
+          }
+        | '('query_expression ')'
+          {
+            $$ = $2;
+          }
+        ;
+
+query_term:
+          query_primary
+        ;
```
Comment (dey4ss) [inline src/parser/bison_parser.y:885]: knit-picky: we usually put short declarations on a single line. I also see no reason to have a non-terminal that only forwards to another one.

File: src/parser/bison_parser.y [hunk 7]
```diff
@@ -756,67 +759,137 @@ update_clause : IDENTIFIER '=' expr {
...
+
+subquery:
+          '(' query_expression ')'
+          {
+            $$ = $2;
+          }
+        ;
```
Comment (dey4ss) [inline src/parser/bison_parser.y:892]: We prefer the following style:
```suggestion
subquery: '(' query_expression ')' {
  $$ = $2;
};
```

(or even a one-liner)

File: test/queries/queries-good.sql [hunk 8]
```diff
@@ -18,7 +18,35 @@ SELECT * FROM t WHERE a = ? AND b = ?;
...
+((SELECT 1)) UNION (SELECT 1);
+(((SELECT 1)) UNION (SELECT 1));
+((((SELECT 1)) UNION (SELECT 1)));
+(SELECT 1) UNION ((SELECT 1));
+((SELECT 1) UNION ((SELECT 1)));
+(((SELECT 1) UNION ((SELECT 1))));
+((SELECT 1)) UNION ((SELECT 1));
+(((SELECT 1)) UNION ((SELECT 1)));
+((((SELECT 1)) UNION ((SELECT 1))));
+# JOIN# JOIN
```
Comment (dey4ss) [inline test/queries/queries-good.sql:49]:
```suggestion
# JOIN
```

File: test/queries/queries-good.sql [hunk 9]
```diff
@@ -18,7 +18,35 @@ SELECT * FROM t WHERE a = ? AND b = ?;
 SELECT City.name, Product.category, SUM(price) FROM fact INNER JOIN City ON fact.city_id = City.id INNER JOIN Product ON fact.product_id = Product.id GROUP BY City.name, Product.category;
 SELECT SUBSTR(a, 3, 5) FROM t;
 SELECT * FROM t WHERE a = DATE '1996-12-31';
-# JOIN
+SELECT 1 UNION SELECT 1;
```
Comment (dey4ss) [inline test/queries/queries-good.sql:21]: You added a lot of things that I would love to see tested more properly. Could you please add more complex examples with actual subqueries? I would also be interested in some edge cases that are not on the happy path, e.g., in `queries-bad.sql`. I would also encourage you to think of a thorough test case in `select_tests.cpp`

**Review by dey4ss — COMMENTED** [review]
I just did a quick pass with some easy-to-fix things, e.g., code style. I will need some more time for an in-depth review because the changes are quite heavy. I appreciate that you kept the naming and grammar closer to the SQL standard, which should make it easier to review.

File: Makefile (hunk 1 above)
Comment (dinodork) [reply Makefile:73]: https://en.wikipedia.org/wiki/CFLAGS

TL;DR: CFLAGS/CXXFLAGS is a Makefile standard, for example for a program

```
#include <iostream>

int main() {
#ifdef FOO
  std::cout << "Foo\n";
#endif
}
```
and, using make's default rule:
```
$ CXXFLAGS=-DFOO make -B cflagtest && ./cflagtest 
g++ -DFOO    cflagtest.cpp   -o cflagtest
Foo
```
It's not that common to have hand-written Makefiles call other hand-written Makefiles - you normally let a build system such as CMake or meson do that for you - but when you do, you need to make to sure you follow this standard and pass along CFLAGS/CXXFLAGS.

File: src/parser/bison_parser.y (hunk 3 above)
Comment (dinodork) [reply src/parser/bison_parser.y:66]: https://www.gnu.org/software/bison/manual/html_node/Expect-Decl.html

In other words, bails out with an error if there's a shift/reduce conflict.

File: src/parser/bison_parser.y (hunk 4 above)
Comment (dinodork) [reply src/parser/bison_parser.y:762]: Done.

File: src/parser/bison_parser.y (hunk 5 above)
Comment (dinodork) [reply src/parser/bison_parser.y:814]: *nit-pick ;)

File: src/parser/bison_parser.y (hunk 7 above)
Comment (dinodork) [reply src/parser/bison_parser.y:892]: Done.

File: test/queries/queries-good.sql (hunk 8 above)
Comment (dinodork) [reply test/queries/queries-good.sql:49]: Oops, fixed.

File: test/queries/queries-good.sql (hunk 9 above)
Comment (dinodork) [reply test/queries/queries-good.sql:21]: Good catch, I actually broke union-in-subquery syntax. I added a new commit that addresses it.

File: src/parser/bison_parser.y (hunk 4 above)
Comment (dinodork) [reply src/parser/bison_parser.y:762]: Btw, the formatting script has a bug where it places a space after % in a `%prec` declaration (see last commit)

**Comment by dinodork** [comment]
@dey4ss Conflicts resolve, please feel free to start the CI workflow.

**Comment by dinodork** [comment]
Bump.

**Comment by dinodork** [comment]
Bump #2  @dey4ss

**Comment by Bouncner** [comment]
Hi @dinodork, we currently lack the time to review this PR properly. We would like to merge it, but not without proper reviewing and that might take some more time. Please bear with us.
